Create HOW_TO_CREATE_A_PLAN (#7456)
Expensify/App: HOW_TO_CREATE_A_PLAN.md

## How to create a plan - Contributor version

### Preamble 
If you’re new to working with Expensify, it’s important to know the philosophy behind the design of Expensify’s products and the processes that wrap around them.
- Problems are identified and defined before solutions are proposed
- Features are not added to satisfy personal preferences
- Features are added thoughtfully and minimally

> It’s also important to be familiar with our existing documentation including our [CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) and [README.md](https://github.com/Expensify/App/blob/main/README.md)

Reviewer: ```suggestion
It’s also important to be familiar with our existing documentation including [CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) and [README.md](https://github.com/Expensify/App/blob/main/README.md)
```

## Step 1: Define the problem
What is the problem that needs solving? 
- Be broad
- Define who experiences the problem
- Does this affect the customer experience? How so? 
- How many customers would be affected by this?
- Define why it is a problem

> - Resist the temptation to reverse-engineer a Solution into a matching Problem.  (ie. Problem: We don’t have a car. Solution: Buy a car) 

Reviewer 2: I am just wondering, the direction we should avoid is Solution -> Problem, should the bad example be something like: 

Solution: I want a car.
Problem: We do not have a car.

But this is just coming to my mind given the order of the problem/solution issue. Let me know what you think.

Author: I think you're right cuz it's "reverse-engineer a Solution" which means you would start with the solution to get the problem.. right?  If so, can you make that update?

Reviewer: I like it 👍

> - Resist the temptation to reverse-engineer a Solution into a matching Problem.  (ie. Problem: We don’t have a car. Solution: Buy a car) 

Reviewer 2: ```suggestion
- Resist the temptation to reverse-engineer a Solution into a matching Problem.  (ie. Solution: Let's get a car! Problem: We don’t have a car.) 
```

- There may be more than one problem, so define all of them

> - Keep the language simple - Reduce every word down to its most plain and straightforward manner.  Avoid jargon and, if you use abbreviations, include the full spelling the first time.

Reviewer: I think this is slightly clearer

```suggestion
- Keep the language simple - Reduce every word down to its most plain and straightforward manner.  Avoid jargon. If you use abbreviations, include the full spelling the first time.
```

- Tear apart your statements and have others tear them apart for you until everybody agrees unambiguously as to the nature of the problem. If needed, continue the discussion in #expensify-open-source.

Identify which problems are actually independent and can be solved by themselves, versus which are actually a series of dependent but otherwise separate problems that can be solved one at a time.

> The broader and better contextualized the problem definition, the greater opportunity for discovering radical, out-of-the-box solutions. The odds of achieving something are dramatically improved by clearly defining what it is.  Take at least as much care perfecting your problem definition as you do the solution, because the perfect solution to the wrong problem is typically an awful solution to the right one.

Reviewer 2: I think there are two spaces here. Suggestion
```suggestion
The broader and better contextualized the problem definition, the greater opportunity for discovering radical, out-of-the-box solutions. The odds of achieving something are dramatically improved by clearly defining what it is. Take at least as much care perfecting your problem definition as you do the solution, because the perfect solution to the wrong problem is typically an awful solution to the right one.
```

## Step 2: Workshop the solutions
Draft as many solutions as you can for each problem, including any existing workarounds - “Do nothing” is always a potential solution. Be as creative or audacious as you want when workshopping potential solutions.

Once you have your solutions, it’s time to decide on the preferred solution (alternate solutions may be included at a later stage). Aspects that may promote one solution to “preferred” status:

> - ROI - the current value of the problem, versus the cost of your solution

Reviewer: I guess we should define this since above we say define acronyms :D

```suggestion
- ROI (return on investment) - the current value of the problem, versus the cost of your solution
```

- Work involved - is one solution just as effective but would take half the time to create?
- Adding or moving “things” - does one solution have less consumer impact?
- Future-proofing - does one solution have more longevity or pave the way for future development?
- Independence - does a solution rely on a different problem to be solved first? Does it rely on another piece to be done later?

> If you are finding the solution to be difficult, go back and beat harder on the problem to break it up into smaller pieces.  Keep repeating until you have a general list of prioritized stages, with early stages solving the dependencies required by later stages, each of which is extremely well defined, with a reasonably obvious preferred solution.

Reviewer 2: Two spaces too.
```suggestion
If you are finding the solution to be difficult, go back and beat harder on the problem to break it up into smaller pieces. Keep repeating until you have a general list of prioritized stages, with early stages solving the dependencies required by later stages, each of which is extremely well defined, with a reasonably obvious preferred solution.
```

## Step 3: Write out each problem and solution (P/S statement)

> Have a trusted peer or two proof your P/S statement and help you ensure it is well-defined.  If you're in need of a peer to proof, post in #expensify-open-source to ask for help.   Refine it and then share with another peer or two until you have a clear, understandable P/S statement. The more complex the problem and solution, the more people should review it. Keep going back to step 1 if needed.

Reviewer 2: Also more spaces.
```suggestion
Have a trusted peer or two proof your P/S statement and help you ensure it is well-defined. If you're in need of a peer to proof, post in #expensify-open-source to ask for help. Refine it and then share with another peer or two until you have a clear, understandable P/S statement. The more complex the problem and solution, the more people should review it. Keep going back to step 1 if needed.
```

## Step 4: Propose it
Once you have a well-defined problem and solution, post your problem and solution as separate comments in ​​#expensify-open-source in Slack and be prepared to answer questions and defend your choices. Also, be prepared to hear better solutions that may completely change your P/S statement. 

> Depending on the complexity of the P/S Statement, a Design Doc may be required.

Reviewer 2: @mallenexpensify Adding the suggestion in that case. 
```suggestion
Depending on the complexity of the P/S Statement, an internal Design Doc may be required.
```

Reviewer 2: @Beamanator I have updated this suggestion, with your suggestion, to form our mutual suggestion! 😄

> Depending on the complexity of the P/S Statement, a Design Doc may be required.

Reviewer 2: I think we should also explain what is a Design doc or link to some explanation.

Author: Contributors won't have the context of a design doc so I don't want to go into too much detail yet.  What about updating to "an internal Design Doc may be required." to stress it's not something a contributor should be looking for?

> link to some explanation

I think if add 'internal' we don't need more detail. But.. I initially read it as 'link to some examples'.  Once we have a few good examples in #expensify-open-source we should add links to them.  Or... create a new .md file with a handful of examples (since not everyone will be in #expensify-open-source)

Reviewer 2: Agreed! You are right, it is internal, but it might be good to imply that they basically _do not have to know what it is_, kind of.

Reviewer: What if we add `(Expensify employees will take it from there)` to make it clear that at that point it's out of the contributor's hands?
